pkg/controllers/consolidation/helpers.go

```diff
@@ -50,10 +55,11 @@ func GetPodEvictionCost(ctx context.Context, p *v1.Pod) float64 {
 	return clamp(-10.0, cost, 10.0)
 }
 
-func filterByPrice(options []cloudprovider.InstanceType, price float64, inclusive bool) []cloudprovider.InstanceType {
+func filterByPrice(options []cloudprovider.InstanceType, reqs scheduling.Requirements, price float64) []cloudprovider.InstanceType {
 	var result []cloudprovider.InstanceType
 	for _, it := range options {
-		if (it.Price() < price) || (inclusive && it.Price() == price) {
+		cheapestOffering := cputils.CheapestOfferingWithReqs(cputils.AvailableOfferings(it), reqs)
```

Comment:
I don't think this is correct as we aren't guaranteed to launch the cheapest offering here, but the problem is pretty rare since we don't do spot->spot replacement.  However I have seen a spot price be equal to the on-demand price which makes this scenario possible

Suppose you have an instance that is an On-Demand at 2.00 and offerings:
- Spot/Zone-A  1.00
- Spot/Zone-B  2.00

this code would consider the instance type as a valid replacement as the offering Spot/Zone-A is cheaper.  However since there's no requirement that gets imposed on the node request, CreateFleet could technically still launch a Spot/Zone-B at the same price which won't save money but is a downgrade as the node went from on-demand to spot at the same price.

I think this bug exists in the original launch code as well as we would prefer to launch the spot node and should prefer to launch the on-demand since it costs the same. It also occurs where we have multiple instance types that can suffice and we launch a more expensive spot node than the cheapest on-demand node would have been.  I've been meaning to fix that for a while.

Reply:
Sure, but I'm not sure the alternative is much better where we average out the offerings price or we assume the worst-case and take the most expensive spot offering.

Perhaps we filter out the offerings here too so that we only end up passing through the offerings that are lower in price than the current offering?

Reply:
Also, the same problem exists in scheduling; however, on some level, we have to take some heuristic to order the instance types and determine what is a "cheaper" instance type. This is; however, less of a problem with initial scheduling since we are making no claim about us launching a node making the nodeset cheaper overall

Reply:
I think assuming the max price is safer, but it wouldn't do an on-demand to spot node replacement.  It seems like we really need to add a requirement to the node that forces the cheapest offering to be selected, but I'm a bit worried of the implications of that.  We're removing some flexibility from CreateFleet due to up to 12-hour old pricing information.

Reply:
For some more information, the scenario I want to avoid is a second cloud provider comes along and consolidation gets in a loop repeatedly trying to replace a node with cheaper instance which continually fails since the cheaper instance doesn't launch. I don't think it will happen for us currently due to spot to spot replacement avoidance and on-demand prices be >= spot prices.

Reply:
It seems like filtering the offerings may solve the issue here since we won't generate any launch template overrides for capacity type/zone that don't appear in the offerings for an instance type, meaning that if we filter them away from the instance types given to the nodeTemplate, we will not be producing instances that wouldn't satisfy our pricing requirement.